handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,203 @@
+    private static final long MIN_TIMEOUT_NANOS = TimeUnit.MILLISECONDS.toNanos(1);
+
+    private final long timeoutNanos;
+    private final boolean closeOnTimeout;
+
+    //shared state, consider removing
+    private ChannelFutureListener channelFutureListener;
+    private ReadTimeoutTask readTimeoutTask;
+
+    private long readFinishTime;
+    private long writeFinishTime;
+    private volatile boolean removed;
+    private boolean closed;
+    private ScheduledFuture<?> timeout;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout after last write to the channel
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     * @param closeOnTimeout if channel should be closed by this handler after firing {@link ReadTimeoutException}
+     *
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit, boolean closeOnTimeout) {
+        if (timeout <= 0) {
+            timeoutNanos = 0;
+        } else {
+            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
+        }
+        this.closeOnTimeout = closeOnTimeout;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true);
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        init(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelFutureListener));
+    }
+
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
+     * timeout cancelling. Default implementation returns {@code true}.
+     * @param ctx channel handler context of related {@link #channelRead} invocation
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(ChannelHandlerContext ctx, Object msg) {
+        return true;
+    }
+
+    private void init(final ChannelHandlerContext ctx) {
```

Comment:
seems like this code is only called in `handlerAdded`... Why not just move in there and remove all the null checks ?